server/src/main/java/org/elasticsearch/repositories/RepositoryData.java

```diff
@@ -253,23 +261,24 @@ public RepositoryData removeSnapshot(final SnapshotId snapshotId, final ShardGen
         newSnapshotStates.remove(snapshotId.getUUID());
         final Map<String, Version> newSnapshotVersions = new HashMap<>(snapshotVersions);
         newSnapshotVersions.remove(snapshotId.getUUID());
-        Map<IndexId, Set<SnapshotId>> indexSnapshots = new HashMap<>();
+        Map<IndexId, List<SnapshotId>> indexSnapshots = new HashMap<>();
         for (final IndexId indexId : indices.values()) {
-            Set<SnapshotId> set;
-            Set<SnapshotId> snapshotIds = this.indexSnapshots.get(indexId);
+            List<SnapshotId> remaining;
+            List<SnapshotId> snapshotIds = this.indexSnapshots.get(indexId);
             assert snapshotIds != null;
             if (snapshotIds.contains(snapshotId)) {
```

Comment:
Not great that we're quadratic here now (for the nested loop), but I don't think it really matters much relative to the significant space+GC savings.